### PR TITLE
Callback promises migration

### PR DESCRIPTION
**Pull request by ashah-splunk**
- Callbacks have been removed and instead we are returning Promises which enables users to use Async/Await features of JS.
- Refactored tests cases and SDK lib files with latest JS features.
- added '`response_timeout`'  parameter which enables user to specify the timeout for a particular API call.
- Removed Async.js file and the required methods have been migrated to Utils.js following the Promise structure.

### CODE REVIEW

**Review by vupadhyay — APPROVED** [review]

File: lib/platform/client/jquery_http.js
Comment (fantavlik) [inline lib/platform/client/jquery_http.js:44]: Can you explain a bit about this header? It's not obvious what this is doing, although I see that the cli parses this to set its parameters but I'm not sure why that would be needed since jquery should be the http client that ulimately needs to respect the timeout, I'm not sure why the cli would need this.

File: lib/platform/client/jquery_http.js
Comment (fantavlik) [inline lib/platform/client/jquery_http.js:67]: Can you help me understand this logic? So I see elsewhere in our code we were setting the status to 600 if the request didn't complete (no response). Is that the case being handled here? It feels like now might be a good time to fix this if so, hardcoding an invalid 600 code is probably not something we want to continue doing. I'm open to suggestions on how to address this.

File: lib/platform/client/proxy_http.js
Comment (fantavlik) [inline lib/platform/client/proxy_http.js:90]: same comment as above, not sure this header makes sense to me

Comment (akaila-splunk) [reply lib/platform/client/proxy_http.js:90]: From proxy_http.js we are passing data to the cli.js through which we communicate with apis using needle. we have added this header to make client side code work similar as backend where we replaced abort method with response_timeout parameter(node_http.js file).

File: lib/platform/client/jquery_http.js
Comment (akaila-splunk) [reply lib/platform/client/jquery_http.js:67]: To answer your question, yes here we are handling the response status set to 600 within our code. We will think on other options for handling the 'abort' use case without hardcoding the status to 600 and let you know our findings.

File: lib/platform/client/jquery_http.js
Comment (akaila-splunk) [reply lib/platform/client/jquery_http.js:44]: To keep the code in sync with the backend we had added this header here, but jQuery timeout functionality works differently, so we are thinking of an approach that can be used along with the new abort flow (i.e replacing 600 response status)

Comment (fantavlik) [reply lib/platform/client/jquery_http.js:44]: OK great that sounds good 👍

File: lib/platform/client/proxy_http.js
Comment (fantavlik) [inline lib/platform/client/proxy_http.js:175]: I actually would suggest that we pick a reasonable timeout other than 0 but encourage users to set this themselves - I think needle and jquery probably handle this differently so it would be good to standardize on one, maybe 30 seconds? Leaving long running connections open on the splunkd side could definitely lead to performance issues - browsers will have different timeout defaults that also interact with this but I think consistency across clients would be best. I realize we haven't been doing this on the enterprise SDK side but the cloud SDKs took this approach. We should also call out the ability to set timeouts in our README as well as the default value that will be used if none is provided. What do you think?

File: lib/platform/client/proxy_http.js
Comment (fantavlik) [reply lib/platform/client/proxy_http.js:90]: Interesting, I think I need to dig in more as to how the cli and this client interact

File: lib/platform/client/jquery_http.js
Comment (fantavlik) [reply lib/platform/client/jquery_http.js:44]: I also noticed since the abort flow has changed we no longer support this advertised feature: https://github.com/splunk/splunk-sdk-javascript/blob/master/CHANGELOG.md#you-can-now-abort-asynchronous-http-requests so I think we need to address this in the docs. For jquery it appears that `abort()` is already supported:
```
let jqxhr = $.ajax( ... );
jqxhr.abort();
```
for needle it looks like the returned object is [advertised as a Readable stream ](https://github.com/tomas/needle#needlemethod-url-data-options-callback--20x) but testing in node it also seems that both `.abort()` and `.destroy()` are supported so maybe we could point users to one or both of those?
```node
let req = needle.get('http://www.google.com', function(error, response) { console.log(`err: ${error} response: ${response}`); });
> req.request.aborted
false
> req.abort();
undefined
> req.request.aborted
true
```
can you confirm? if so that's great that the functionality is still available

File: lib/service.js
Comment (fantavlik) [inline lib/service.js:146]: I think this should be `let propsFile = await files.item("props");` does that seem right?

File: lib/service.js
Comment (fantavlik) [inline lib/service.js:4977]: I wouldn't worry about this too much, this would be a new scenario that we weren't handling previously - omitting the response seems perfectly fine to me

File: lib/utils.js
Comment (fantavlik) [inline lib/utils.js:746]: I think these utilities from the Async class are only used by our tests does that seem right? If so maybe we can take an action item after this merges to refactor these out of our tests and remove them

File: lib/platform/client/proxy_http.js
Comment (fantavlik) [inline lib/platform/client/proxy_http.js:150]: This appears to be identical to `makeRequest` above - can we remove it?

Comment (fantavlik) [reply lib/platform/client/proxy_http.js:150]: or was this meant to be `makeRequestAsync: async function(url, message) {` in which case can we have this function call makeRequest?

File: lib/service.js
Comment (ashah-splunk) [reply lib/service.js]: yes, and we have updated it

File: README.md
Comment (ashah-splunk) [reply README.md]: we updated the examples that are part of the docs but missed within the README file. we have updated the examples .

File: lib/platform/client/proxy_http.js
Comment (ashah-splunk) [reply lib/platform/client/proxy_http.js:150]: we had added this method for `conf` CRUD functionality and it was implemented using Promises. Currently we did not merge this with the updated makeRequest method as it would require some analysis and testing on the working conf CRUD feature. we thought to include it as a follow-up once Promise changes are merged.

File: lib/platform/client/jquery_http.js
Comment (akaila-splunk) [reply lib/platform/client/jquery_http.js]: as per our discussion we have removed the header and now we'll consider the max of `timeout` and `response_timeout` for aborting request.

File: lib/platform/client/jquery_http.js
Comment (akaila-splunk) [reply lib/platform/client/jquery_http.js]: we have refactor the code such that we don't require the 600 statuscode.

File: lib/platform/client/proxy_http.js
Comment (akaila-splunk) [reply lib/platform/client/proxy_http.js]: as per our discussion we have removed the header and now we'll consider the max of `timeout` and `response_timeout` for aborting request.

File: lib/utils.js
Comment (ashah-splunk) [reply lib/utils.js:746]: as per our discussion regarding scenarios where users might need to execute some action multiple times such as create/delete users or roles or any resource, these utilities would be useful, hence keeping them as is

File: README.md
Comment (akaila-splunk) [reply README.md]: we have updated the examples in README file.

**Review by fantavlik — APPROVED** [review]
This is looking great, thanks for addressing all of those comments 🚀

File: lib/platform/client/proxy_http.js
Comment (akaila-splunk) [reply lib/platform/client/proxy_http.js]: as discussed we'll create a follow up task for default timeout.

**Pull request merged**
